test/phabricator_test/test_phabricator.py

```diff
@@ -0,0 +1,356 @@
+
+    # TODO: Very similar methods (user_query_ids, user_query_usernames etc. That will be based off 'test_differential_query'.
+
+    @patch(PATH + 'PhabricatorService.user_query_ids')
+    def test_author_data_in_rawReponse(self, mock_user_query_ids):
+        """
+            Tests author_data function where author_phid IS IN the raw response
+        """
+
+        fake_user, fake_raw_response = self.mock.mock_phabricator_author_data_rawResponse()
+
+        user, raw_response = PhabricatorService().author_data(
+            author_phid=fake_user['phid'],
+            raw_response=fake_raw_response,
+            phab=None
+        )
+
+        mock_user_query_ids.assert_not_called()
+        self.assertEqual(fake_user, user)
+        self.assertEqual(fake_raw_response, raw_response)
+
+    @patch(PATH + 'PhabricatorService.user_query_ids',
+           return_value=["New User!"])
+    def test_author_data_not_in_rawResponse(self, mock_user_query_ids):
+        """
+            Tests author_data function where author_phid IS NOT IN the raw response
+        """
+        expected_raw_response = ["New User!"]
+        expected_user = "New User!"
+
+        fake_user, _ = self.mock.mock_phabricator_author_data_rawResponse()
+        
+        user, raw_response = PhabricatorService().author_data(
+            author_phid=fake_user['phid'],
+            raw_response=[],
+            phab=None
+        )
+
+        mock_user_query_ids.assert_called_with(
+            [fake_user['phid']],
+            None
+        )
+        self.assertEqual(expected_user, user)
+        self.assertEqual(expected_raw_response, raw_response)
+
+    @patch(PATH + 'PhabricatorService.user_query_ids',
+           side_effect=Exception('404 Error Exception'))
+    def test_author_data_failure(self, mock_user_query_ids):
+        """
+            Tests author_data function where user_query_ids throws an error (i.e. 404)
+        """
+
+        fake_user, _ = self.mock.mock_phabricator_author_data_rawResponse()
+
+        with self.assertRaises(Exception) as context:
+            user, raw_response = PhabricatorService().author_data(
+                author_phid=fake_user['phid'],
+                raw_response=[],
+                phab=None
+            )
+
+        self.assertIn('404 Error Exception', context.exception)
+        mock_user_query_ids.assert_called_with(
+            [fake_user['phid']],
+            None
+        )
+
+    @patch(PATH + 'PhabricatorService.author_data',
+           return_value=({'userName': 'dummy-user'}, []))
+    @patch(PATH + 'PhabricatorService.time_from_epoch',
+           return_value= 'dummy-time')
+    def test_get_last_comment_successful_single(self, mock_time_from_epoch,
+                                                mock_author_data):
+        """
+            Tests get_last_comment function (successful, fake_comments have one comment)
+        """
+
+        expected_response = LastComment(author='dummy-user',
```

Comment:
Remember that the method should be completely isolated. So you should mock the LastComment in the `phabricatorstack.py` that returns some fake comment, and then verify that returned value from the function is that fake comment + check that the mocked LastComment got expected parameters.

Reply:
LastComment is just a named tuple. See: 

`LastComment = namedtuple('LastComment', ('author', 'body', 'created_at'))`

And not a class. Would I still have to mock it even though its a data structure? I'm a bit confused about how I could mock it.

Reply:
As LastComment is a named tuple object: 

`LastComment = namedtuple('LastComment', ('author', 'body', 'created_at'))`

I'm having a bit of trouble mocking it: 

`     mock_last_comment().assert_called_with(
             author='dummy-user',
             body='This is some content',
             created_at='dummy-time')
`

Trying to assert what the function was called with results in: 

`AttributeError: 'LastComment' object has no attribute 'assert_called_with'`

Reply:
The issue lies in the path you are mocking. If you import LastComment in `reviewrot.phabricatorstack`, the object full path is `reviewrot.phabricatorstack.LastComment` instead of `reviewrot.basereview.LastComment` (because the object `LastComment` is imported into the namespace `reviewrot.phabricatorstack`). This is basically how python uses namespaces and structure the variables etc.

Reply:
This works for me
```python
    @patch(PATH + 'PhabricatorService.author_data',
           return_value=({'userName': 'dummy-user'}, []))
    @patch(PATH + 'PhabricatorService.time_from_epoch',
           return_value= 'dummy-time')
    @patch(PATH + 'LastComment')
    def test_get_last_comment_successful_single(self, mock_last_comment,
                                                mock_time_from_epoch,
                                                mock_author_data):
        """
        Tests get_last_comment function (successful, fake_comments have one comment)
        """
        mock_last_comment.return_value = MagicMock()

        fake_comments = self.mock.mock_phabricator_get_comments_single()
        response = PhabricatorService().get_last_comment(
            comments=fake_comments,
            phab=None,
            raw_response=[]
        )
        mock_author_data.assert_called_with(
            author_phid='PHID-USER-xxxxxxxxxxxxxxxxxxxx',
            phab=None,
            raw_response=[]
        )
        mock_time_from_epoch.assert_called_with('1551763640')
        mock_last_comment.assert_called_with(
            author='dummy-user',
            body='This is some content',
            created_at='dummy-time'
        )
        self.assertEqual(response, mock_last_comment.return_value)
```

Additionally, it would be better not to use LastComment and create something I did instead because in a case that something will get wrong with LastComment (let's say when you create an instance, it will raise some KeyError or something), you get false report that method get_last_comment is wrong as well.